File: examples/javascript/rest/valuation.mjs
/**
 * Copyright 2026 Volstrata.com
 *
 * The `valuation.*` and `fundamentals.*` domains: a screened universe, one
 * company's valuation record, its fundamentals, and the one POST capability in
 * this file — which needs the Pro plan.
 * Catalog: https://volstrata.com/docs/api-catalog
 *
 * Run:  node rest/valuation.mjs [TICKER]
 *       npm run valuation
 *
 * Four requests. Unlike the index-oriented domains, these take an equity ticker,
 * so the default here is AAPL rather than SPX.
 *
 *   valuation.universe    GET   needs a credential
 *   valuation.company     GET   needs a credential
 *   fundamentals.company  GET   answers anonymously
 *   valuation.value       POST  needs the Pro plan — the shape of a POST call
 *
 * Worth knowing before you run this with no key: the `valuation.*` rows answer
 * 401 `auth_required` to an anonymous caller, so three of the four rows below
 * refuse. That is the point of running it — a credential gets you past a 401,
 * and whether the data then comes back is a separate question your plan
 * answers. `GET /api/v1/meta/access` reports what your own credential reaches.
 */

import {
  apiGet,
  apiPost,
  banner,
  callCapability,
  heading,
  keysOf,
  report,
  show,
  summarize,
} from "../lib/volstrata.mjs";

const ticker = (process.argv[2] ?? "AAPL").toUpperCase();

banner(`valuation.* — ${ticker}`);
show("ticker", `${ticker} (pass another as the first argument)`);

const results = [];

// ---------------------------------------------------------------------------
// valuation.universe — what is covered at all
// ---------------------------------------------------------------------------
//
// Start here rather than guessing whether a symbol is covered. The listing takes
// `limit` and `offset` (this one is offset-paged, not cursor-paged — the shared
// cursor model in pagination.mjs applies to the 17 operations that declare a
// `cursor` parameter, and this is not one of them), plus optional `sector`,
// `model`, `verdict` and `min_market_cap` filters.

results.push(
  await callCapability("valuation.universe", async () => {
    const data = await apiGet("/api/v1/valuation/universe", { limit: 5, offset: 0 });

    show("keys", keysOf(data));
    const [rowsKey, rows] = Object.entries(data).find(([, v]) => Array.isArray(v)) ?? [];
    if (rows) {
      show("rows key", rowsKey);
      show("rows returned", rows.length);
      for (const row of rows.slice(0, 3)) {
        console.log(`      ${summarize(row, { depth: 1, maxKeys: 7 })}`);
      }
    } else {
      show("payload", summarize(data, { depth: 2, maxKeys: 12 }));
    }
    return data;
  }, { note: "limit=5 — keep exploratory calls small" }),
);

// ---------------------------------------------------------------------------
// valuation.company — one company's record
// ---------------------------------------------------------------------------
//
// The response fields are the outputs of a valuation model. This repo documents
// their names and units and links the glossary for what each one means; how any
// of them is produced is not something these examples describe.

results.push(
  await callCapability("valuation.company", async () => {
    const data = await apiGet("/api/v1/valuation/company", { ticker });
    show("keys", keysOf(data, 20));
    show("payload", summarize(data, { depth: 2, maxKeys: 10 }));
    return data;
  }),
);

// ---------------------------------------------------------------------------
// fundamentals.company — the reported figures behind it
// ---------------------------------------------------------------------------

results.push(
  await callCapability("fundamentals.company", async () => {
    const data = await apiGet("/api/v1/fundamentals/company", { ticker });
    show("keys", keysOf(data, 20));
    show("payload", summarize(data, { depth: 2, maxKeys: 10 }));
    return data;
  }),
);

// ---------------------------------------------------------------------------
// valuation.value — POST, Pro
// ---------------------------------------------------------------------------
//
// 7 of the 180 published operations are POST; the other 173 are GET. A POST
// capability takes a JSON body:
//
//   fetch(url, {
//     method: "POST",
//     headers: { "Content-Type": "application/json", ... },
//     body: JSON.stringify({ ticker: "AAPL" }),
//   })
//
// which is what `apiPost` below does. Everything else is identical to a GET:
// the same Authorization header, the same problem+json refusal, the same
// rate-limit headers. Without the Pro plan this row refuses and the run
// continues.

results.push(
  await callCapability("valuation.value", async () => {
    const data = await apiPost("/api/v1/valuation/value", { ticker });
    show("keys", keysOf(data, 20));
    show("payload", summarize(data, { depth: 2, maxKeys: 10 }));
    return data;
  }, { note: "Pro plan — POST with a JSON body" }),
);

report(results);

heading("Related");
console.log("  Capability catalog   https://volstrata.com/docs/api-catalog");
console.log("  Plans and auth       https://volstrata.com/docs/api-auth");
console.log("  Next domain          node rest/market.mjs");

File: examples/python/rest/valuation.py
"""Copyright 2026 Volstrata.com

The `valuation` and `fundamentals` domains, and the first POST in this repo.
Docs: https://volstrata.com/docs/api-catalog
      https://volstrata.com/docs/api-auth

    python rest/valuation.py

Four capabilities:

    fundamentals.company  GET  /api/v1/fundamentals/company  reported figures
    valuation.universe    GET  /api/v1/valuation/universe    a screen, paged by offset
    valuation.company     GET  /api/v1/valuation/company     one company
    valuation.value       POST /api/v1/valuation/value       above the lowest floor

Two things this file is here to teach:

  * A capability whose catalog floor is Free can still require an authenticated
    principal. When that is the case the refusal is a 401 `auth_required` rather
    than a 403 naming a plan -- any accepted credential clears it. Check
    GET /api/v1/meta/access for what your own credential unlocks.
  * `valuation.universe` pages with `limit` and `offset`, not with a cursor.
    Most of the surface is cursor-paged; this one is not, and reading the
    parameters in the spec rather than assuming is the habit that catches it.

Four sequential requests. Some will be refused unless you have a key.
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from volstrata_helpers import (  # noqa: E402  (import after the path fix, on purpose)
    attempt,
    get,
    heading,
    key_banner,
    post,
    row,
    run_examples,
    show,
)

TICKER = "AAPL"


def reported_fundamentals() -> None:
    """fundamentals.company returns the reported line items, plus their history.

    Shape:

        {"ok": true,
         "entity_name": "...", "cik": "...", "currency": "USD",
         "fiscal_year": <year>, "period_end": "YYYY-MM-DD", "as_of": <unix>,
         "latest":  {"revenue": <n>, "net_income": <n>, "eps_diluted": <n>, ...},
         "history": {"revenue": [{"fy": <year>, "end": "YYYY-MM-DD", "val": <n>}, ...],
                      ...}}

    `latest` and `history` share their keys, so one loop can render a value and
    its trend. Amounts are in `currency` units as reported; per-share figures are
    per share. Nothing here is derived or adjusted -- it is what was filed.
    """
    heading("fundamentals.company -- reported figures")
    data = get("/fundamentals/company", ticker=TICKER)
    row("entity_name", data.get("entity_name"))
    row("cik", data.get("cik"))
    row("currency", data.get("currency"))
    row("fiscal_year", data.get("fiscal_year"))
    row("period_end", data.get("period_end"))

    latest = data.get("latest") or {}
    row("line items", len(latest))
    for field in ("revenue", "net_income", "eps_diluted", "total_assets", "cash"):
        if field in latest:
            row(field, latest[field])

    history = data.get("history") or {}
    revenue = history.get("revenue") or []
    if revenue:
        # Each point is {"fy": <year>, "end": "YYYY-MM-DD", "val": <n>}. Sort by
        # `end` yourself if order matters to you rather than trusting the order
        # a list happens to arrive in.
        print("  history.revenue, first rows as returned:")
        for point in revenue[:4]:
            print("    fy {0}  ending {1}".format(point.get("fy"), point.get("end")))


def screen_the_universe() -> None:
    """valuation.universe is a screen: many companies, a few fields each.

    Parameters: limit, offset, sector, model, verdict, min_market_cap, format.
    Page it by advancing `offset` -- and note the ceiling on `limit` is the
    server's to choose, so read how many rows you actually got rather than
    assuming you got `limit` of them.
    """
    heading("valuation.universe -- a screen")
    data = attempt("valuation.universe", get, "/valuation/universe", limit=5, offset=0)
    if data is None:
        print("  Set VOLSTRATA_API_KEY and re-run to see this one.")
        return
    rows = None
    for key in ("universe", "rows", "companies", "results"):
        if isinstance(data.get(key), list):
            rows = data[key]
            row("collection key", key)
            break
    row("rows returned", len(rows) if rows is not None else "n/a")
    if rows:
        show(rows[0])
    else:
        # Not every response names its collection the same way; when in doubt,
        # print the keys and read them once rather than guessing forever.
        row("top-level keys", ", ".join(sorted(data)[:14]))


def one_company() -> None:
    """valuation.company is the single-name view of the same data set."""
    heading("valuation.company -- one company")
    data = attempt("valuation.company", get, "/valuation/company", ticker=TICKER)
    if data is None:
        print("  Same credential requirement as the screen above.")
        return
    row("ticker", data.get("ticker"))
    row("top-level keys", ", ".join(sorted(data)[:14]))
    show(data, max_items=3, max_chars=900)


def gated_post() -> None:
    """valuation.value is a POST, and it is above the lowest floor.

    Seven of the 180 published operations take POST; the rest are GET. The body
    is plain JSON, and `requests` will set Content-Type for you when you pass
    `json=`:

        POST /api/v1/valuation/value
        Content-Type: application/json

        {"ticker": "AAPL", "format": "json"}

    The `overrides` parameter is documented in the spec for callers who want to
    supply their own assumptions. The refusal below names the plan that clears
    the floor; nothing about the request shape is wrong.
    """
    heading("valuation.value -- gated POST")
    body = {"ticker": TICKER, "format": "json"}
    print("  body: {0}".format(body))
    data = attempt("valuation.value", post, "/valuation/value", body)
    if data is None:
        print("  Expected without the plan. Pricing and keys:")
        print("  https://volstrata.com/api-keys")
        return
    row("ok", data.get("ok"))
    row("top-level keys", ", ".join(sorted(data)[:14]))


def main() -> int:
    print("VolStrata: valuation and fundamentals -- https://volstrata.com")
    key_banner()
    exit_code = run_examples((
        ("fundamentals", reported_fundamentals),
        ("universe", screen_the_universe),
        ("company", one_company),
        ("valuation.value", gated_post),
    ))
    heading("Next")
    print("  discover_capabilities.py   what your credential unlocks")
    print("  rest/research_and_docs.py  the research query surface")
    print("  Catalog: https://volstrata.com/docs/api-catalog")
    return exit_code


if __name__ == "__main__":
    sys.exit(main())
